src/libraries/System.Private.CoreLib/src/System/Random.cs

```diff
@@ -1,197 +1,599 @@
+
+        /// <summary>Base type for all generator implementations that plug into the base Random.</summary>
+        private abstract class ImplBase
         {
-            if (minValue > maxValue)
-            {
-                throw new ArgumentOutOfRangeException(nameof(minValue), SR.Format(SR.Argument_MinMaxValue, nameof(minValue), nameof(maxValue)));
-            }
+            public abstract double Sample();
 
-            long range = (long)maxValue - minValue;
-            return range <= int.MaxValue ?
-                (int)(Sample() * range) + minValue :
-                (int)((long)(GetSampleForLargeRange() * range) + minValue);
-        }
+            public abstract int Next();
 
-        public virtual double NextDouble() => Sample();
+            public abstract int Next(int maxValue);
 
-        public virtual void NextBytes(byte[] buffer)
+            public abstract int Next(int minValue, int maxValue);
+
+            public abstract long NextInt64();
+
+            public abstract long NextInt64(long maxValue);
+
+            public abstract long NextInt64(long minValue, long maxValue);
+
+            public abstract double NextDouble();
+
+            public abstract float NextSingle();
+
+            public abstract void NextBytes(byte[] buffer);
+
+            public abstract void NextBytes(Span<byte> buffer);
+        }
+
+        #region Implementations
+        /// <summary>
+        /// Provides an implementation of the xoshiro256** algorithm. This implementation is used
+        /// when no seed is specified and an instance of the base Randomclass is constructed.  As
+        /// such, we are free to implement however we see fit, withoutback compat concerns around
+        /// the sequence of numbers generated or what methods call what other methods.
+        /// </summary>
+        private sealed class Xoshiro256StarStarImpl : ImplBase
         {
-            if (buffer is null)
+            // NextUInt64 is based on the algorithm from http://prng.di.unimi.it/xoshiro256starstar.c:
+            //
+            //     Written in 2018 by David Blackman and Sebastiano Vigna (vigna@acm.org)
+            //
+            //     To the extent possible under law, the author has dedicated all copyright
+            //     and related and neighboring rights to this software to the public domain
+            //     worldwide. This software is distributed without any warranty.
+            //
+            //     See <http://creativecommons.org/publicdomain/zero/1.0/>.
+
+            private ulong _s0, _s1, _s2, _s3;
+
+            public unsafe Xoshiro256StarStarImpl()
             {
-                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.buffer);
+                do
+                {
+                    Guid g;
+
+                    // Seed the four ulongs (4 64-bit values) from two Guids (2 128-bit values)
+                    g = Guid.NewGuid();
```

Comment:
This is not using thread-static Random anymore. Is it going to regress https://github.com/dotnet/runtime/commit/6ff058483546657e5a85e1f2e1a35fc11ee5d27a ?

Reply:
I don't think so: that problem was caused by seeding with the tick count, which didn't tick fast enough. There should be no such problem with generating Guids, nor if we change to cryptographically random seeding.

We may need the thread static perf trick from that commit, if the perf tests on Linux show the constructor is slow because Guids/entropy is slower there. But it wouldn't be a correctness issue.
IIRC, new guid on Linux can be significantly slower, if that turns out to be the case then it may be necessary to use thread static as before so it is only necessary for the very first one, and after that they are seeded using that special one.

Reply:
> This is not using thread-static Random anymore. Is it going to regress 6ff0584 ?

Having measured now on both Windows and Linux:

Windows 64-bit

|        Method |          Toolchain |         Mean | Ratio | Allocated |
|-------------- |------------------- |-------------:|------:|----------:|
|          Ctor | master\corerun.exe | 1,454.062 ns |  1.00 |     280 B |
|          Ctor |     pr\corerun.exe |   112.203 ns |  0.08 |      72 B |

Windows 32-bit

|        Method |              Toolchain |         Mean | Ratio | Allocated |
|-------------- |----------------------- |-------------:|------:|----------:|
|          Ctor | master_x86\corerun.exe | 1,680.583 ns |  1.00 |     256 B |
|          Ctor |     pr_x86\corerun.exe |   259.524 ns |  0.15 |      36 B |

Linux 64-bit

|        Method |       Toolchain |         Mean | Ratio | Allocated |
|-------------- |---------------- |-------------:|------:|----------:|
|          Ctor | /master/corerun | 1,510.009 ns |  1.00 |     280 B |
|          Ctor |     /pr/corerun |   783.730 ns |  0.52 |      72 B |

So at least on my machines, there's no regression in the Ctor, and on the contrary, all of them get significantly faster.  However, as shown, the improvement on Windows is larger than the improvement on Linux, because Interop.GetRandomBytes is slower on Linux than on Windows.  If subsequently we want to improve that and we can't improve it in Interop.GetRandomBytes itself, we could play tricks in Random with a thread-static seed generator instance again in the default implementation.